### PR TITLE
Implement SIP 64 as non-experimental

### PR DESCRIPTION
**Pull request by odersky**
For the new syntax of givens and context bounds, see the diffs in reference/syntax.md.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/config/SourceVersion.scala
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/config/SourceVersion.scala:19]: should add this to the `illegalSourceVersionNames` list so users can't choose it with `-source:never`

File: docs/_docs/reference/contextual/context-bounds.md
Comment (bishabosha) [inline docs/_docs/reference/contextual/context-bounds.md:147]:
```suggestion
val less: Comparer = [X] => (x: X, y: X) => (ord: Ord[X]) ?=>
```

**Comment by odersky** [comment]
Who would like to gve this a review? It's all pretty straightforward. Biggest real changes are in the docs.

File: compiler/src/dotty/tools/dotc/config/SourceVersion.scala
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/config/SourceVersion.scala:37]: I found that also this needs changing (the available choices to the `-source` flag:
https://github.com/scala/scala3/blob/56a384fd118c4a79218fd76e7ee2bb0cc12c83e3/compiler/src/dotty/tools/dotc/config/ScalaSettingsProperties.scala#L30-L31

Comment (odersky) [reply compiler/src/dotty/tools/dotc/config/SourceVersion.scala:37]: I have rolled that change into another PR (the named tuples one). 

Do you have time to do a full review by the end of the week? If we want to get this into 3.6 we need to be done by then.

**Comment by smarter** [comment]
Playing a bit with the implementation, I noticed some details that I believe weren't clear from the spec:

## Context bounds for type members

I love this feature but I wonder if it could be used to bring extension methods in scope too:
```scala
trait Serializable[T]:
  extension (x: T) def serialize: String

trait Base:
  type Elem: Serializable

def test0(b: Base) =
  summon[Serializable[b.Elem]] // ok
  b.serialize // error
```

## Naming context bounds

I expected the `as` bounds on classes and traits to create public given members:

```scala
case class CC1[X: Serializable as s](x: X):
  type Elem = X

def test1(c: CC1[?]) =
  c.s // error
  summon[Serializable[c.Elem]] // error
```

Adding a public alias isn't enough to have it summon-able like in the context bounds for type members example above:
```scala
case class CC2[X: Serializable as s](x: X):
  type Elem = X
  given s1: Serializable[Elem] = s

def test2(c: CC2[?]) =
  c.s1 // ok
  summon[Serializable[c.Elem]] // error
```

Even if we directly extend from a trait defining a context bound for a type member, I guess because of dealiasing:
```scala
case class CC3[X: Serializable](x: X) extends Base:
  type Elem = X

def test3(c: CC3[?]) =
  summon[Serializable[c.Elem]] // error
  val b: Base = c
  summon[Serializable[b.Elem]] // ok
```

<hr>

Ultimately, I wish member selection worked as well for regular type bounds than context bounds:
```scala
case class Foo[T <: Serializer](x: T)
Foo(1).serialize
// vs
case class Foo[T: Serializable](x: T) // maybe with `as s` so I still have the ability to not retain a reference to the typeclass instance
Foo(1).serialize
```

I don't see a clear path to get there so this is more food for thought than a PR review.

**Comment by odersky** [comment]
@smarter I think these additions would be good to have as feature requests.

File: docs/_docs/internals/syntax.md
Comment (kyouko-taiga) [inline docs/_docs/internals/syntax.md:489]: nit: consistent alignment
```suggestion
OldGivenSig       ::=  [id] [DefTypeParamClause] {UsingParamClause} ‘:’          -- one of `id`, `DefTypeParamClause`, `UsingParamClause` must be present
```

File: docs/_docs/reference/contextual/context-bounds.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/context-bounds.md:56]: Perhaps there should be a note saying what happens if one wants to combine a subtype bound with *multiple* context bounds.

File: docs/_docs/reference/contextual/context-bounds.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/context-bounds.md:40]: Perhaps there should be an example showing where the `as` clauses are attached when there are more than one context bound.

File: docs/_docs/reference/contextual/context-bounds.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/context-bounds.md:74]: The phrasing of this sentence sounds odd. Shouldn't it:
```suggestion
 3. Otherwise, the parameters arising from context bounds form a new using clause at the end.
```

File: docs/_docs/reference/contextual/context-bounds.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/context-bounds.md:158]: I think the expansion should be shown.

File: docs/_docs/reference/contextual/context-bounds.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/context-bounds.md:162]:
```suggestion
From Scala 3.6 on, context bounds can be used not only for type parameters but also for abstract type members.
```

File: docs/_docs/reference/contextual/givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/givens.md:21]: Why not using a `as` clause here? Then there'd be no need for the `summon` below.

File: docs/_docs/reference/contextual/givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/givens.md:86]: nit: repetition
```suggestion
time it is accessed. If the given is a mere alias to some immutable value, the given is implemented as a simple forwarder, without incurring the cost of a field to hold a cached value. If a given is conditional, a fresh instance is created for each reference.
```

File: docs/_docs/reference/contextual/givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/givens.md:109]:
```suggestion
A given instance starts with the reserved keyword `given`, which is followed by
```

File: docs/_docs/reference/contextual/givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/givens.md:90]:
```suggestion
Here is the full syntax for given instances. Some of these forms of givens are explained in a separate page: [Other Forms of Givens](../more-givens.md).
```

File: docs/_docs/reference/contextual/more-givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/more-givens.md:48]:
```suggestion
Though in general we want to avoid re-evaluating a given, there are situations where such a re-evaluation may be necessary. For instance, say we have a mutable variable `curCtx` and we want to define a given that returns the current value of that variable. A normal given alias will not do since by default given aliases are mapped to lazy vals. In this case, we can specify a _by-name_ evaluation insteadby writing a conditional given with an empty parameter list:
```

File: docs/_docs/reference/contextual/more-givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/more-givens.md:53]:
```suggestion
With this definition, each time a `Context` is summoned we evaluate the `context` function, which produces the current value of `curCtx`.
```

File: docs/_docs/reference/contextual/previous-givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/previous-givens.md:163]:
```suggestion
```

File: docs/_docs/reference/contextual/previous-givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/previous-givens.md:218]:
```suggestion
                     |   'given' GivenDef
```

File: docs/_docs/reference/contextual/previous-givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/previous-givens.md:220]:
```suggestion
                     |   [GivenSig] AnnotType '=0 Expr
```

File: docs/_docs/reference/contextual/previous-givens.md
Comment (kyouko-taiga) [inline docs/_docs/reference/contextual/previous-givens.md:223]:
```suggestion
GivenSig            ::=  [id] [DefTypeParamClause] {UsingParamClause} ':'
StructuralInstance  ::=  ConstrApp {'with' ConstrApp} [‘with’ TemplateBody]
```

File: docs/_docs/reference/experimental/typeclasses.md
Comment (kyouko-taiga) [inline docs/_docs/reference/experimental/typeclasses.md:21]:
```suggestion
in source version `future` if the additional experimental language import `modularity` is present. For instance, using the following command:
```

File: docs/_docs/reference/syntax.md
Comment (kyouko-taiga) [inline docs/_docs/reference/syntax.md:221]:
```suggestion
TypeAndCtxBounds  ::=  TypeBounds [':' ContextBounds]
ContextBounds     ::=  ContextBound
                    |  ContextBound ':' ContextBounds         -- to be deprecated
                    |  '{' ContextBound {',' ContextBound} '}'
ContextBound      ::=  Type ['as' id]
```

File: docs/_docs/reference/syntax.md
Comment (kyouko-taiga) [inline docs/_docs/reference/syntax.md:444]:
```suggestion
                    |  'given' (GivenDef | OldGivenDef)
```

File: docs/_docs/reference/syntax.md
Comment (kyouko-taiga) [inline docs/_docs/reference/syntax.md:464]:
```suggestion
OldGivenDef       ::=  [OldGivenSig] (AnnotType ['=' Expr] | StructuralInstance) -- syntax up to Scala 3.5, to be deprecated in the future
OldGivenSig       ::=  [id] [DefTypeParamClause] {UsingParamClause} ':'          -- one of `id`, `DefTypeParamClause`, `UsingParamClause` must be present
```

**Review by kyouko-taiga — APPROVED** [review]
LGTM overall. I left a few inline comments for things that I think could be improved before it gets merged, but nothing looks like a blocker to me.

Very nice step forward!

File: docs/_docs/reference/contextual/context-bounds.md
Comment (odersky) [reply docs/_docs/reference/contextual/context-bounds.md:56]: I think it's fine since the two are orthogonal and we cover combination of subtype and context bounds in line 22 before.

File: docs/_docs/reference/contextual/context-bounds.md
Comment (odersky) [reply docs/_docs/reference/contextual/context-bounds.md:40]: We have that a little bit later.

File: docs/_docs/reference/contextual/givens.md
Comment (odersky) [reply docs/_docs/reference/contextual/givens.md:21]: It's because we introduce `as` clauses only on the next page.

File: docs/_docs/reference/contextual/context-bounds.md
Comment (WojciechMazur) [inline docs/_docs/reference/contextual/context-bounds.md:140]: I think this example is either not correct - it does not compile in neither plain Scala 3.6-nightly or with experimental:modularity. 
It fails to parse by the compiler

Comment (odersky) [reply docs/_docs/reference/contextual/context-bounds.md:140]: yes that one is still in the works: https://github.com/scala/scala3/pull/21643

File: docs/_docs/reference/contextual/context-bounds.md
Comment (WojciechMazur) [inline docs/_docs/reference/contextual/context-bounds.md:172]: This example might be incorrect aswell. I'm not sure it it's a bug or expected behaviour, but it fails to compile with 
```scala
[error] No given instance of type Ord[Sorted.this.Element] was found for inferring the implementation of the deferred given instance given_Ord_Element in class Sorted
```

It does however compile when replacing `class Collection` with `trait Collection`

Comment (odersky) [reply docs/_docs/reference/contextual/context-bounds.md:172]: Oh yes, that's a bug. It should be trait.

**Pull request merged**
